add a c++ interface in processGroup to get its backend name

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51394 log newly added construction and runtime stats at randomly selected iterations
* #51386 add stats that can only be collected at runtime
* #51260 add more logging fields that can be set in construction time
* #50637 Create PyTorch DDP logging APIs for applications to use
* **#51066 add a c++ interface in processGroup to get its backend name**

backend name of a processgroup created using distributed_c10d python API is tracked, but there is no good way to track name of a processgroup created using processGroup c++ API. In some cases, knowing backend name of a processGroup is useful, e,g., log the backend name, or write some codes that have dependency on the known backend.

Differential Revision: [D26059769](https://our.internmc.facebook.com/intern/diff/D26059769/)

zhaojuanmao committed:
add a c++ interface in processGroup to get its backend name

backend name of a processgroup created using distributed_c10d python API is tracked, but there is no good way to track name of a processgroup created using processGroup c++ API. In some cases, knowing backend name of a processGroup is useful, e,g., log the backend name, or write some codes that have dependency on the known backend.

Differential Revision: [D26059769](https://our.internmc.facebook.com/intern/diff/D26059769/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add a c++ interface in processGroup to get its backend name"


backend name of a processgroup created using distributed_c10d python API is tracked, but there is no good way to track name of a processgroup created using processGroup c++ API. In some cases, knowing backend name of a processGroup is useful, e,g., log the backend name, or write some codes that have dependency on the known backend.

Differential Revision: [D26059769](https://our.internmc.facebook.com/intern/diff/D26059769/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/ProcessGroup.hpp:
```diff
@@ -42,6 +42,12 @@ enum class OpType : std::uint8_t {
   UNKNOWN = 100,
 };
 
+const std::string GLOO_BACKEND_NAME = "gloo";
+const std::string NCCL_BACKEND_NAME = "nccl";
+const std::string MPI_BACKEND_NAME = "mpi";
+const std::string ROUND_ROBIN_BACKEND_NAME = "round_robin";
+const std::string UNDEFINED = "undefined";
```
nit: Probably putting implementation specific constants in an interface file is not right. Can we just use these in the appropriate files for the implementing class?

zhaojuanmao committed:
Update on "add a c++ interface in processGroup to get its backend name"


backend name of a processgroup created using distributed_c10d python API is tracked, but there is no good way to track name of a processgroup created using processGroup c++ API. In some cases, knowing backend name of a processGroup is useful, e,g., log the backend name, or write some codes that have dependency on the known backend.

Differential Revision: [D26059769](https://our.internmc.facebook.com/intern/diff/D26059769/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add a c++ interface in processGroup to get its backend name"


backend name of a processgroup created using distributed_c10d python API is tracked, but there is no good way to track name of a processgroup created using processGroup c++ API. In some cases, knowing backend name of a processGroup is useful, e,g., log the backend name, or write some codes that have dependency on the known backend.

Differential Revision: [D26059769](https://our.internmc.facebook.com/intern/diff/D26059769/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add a c++ interface in processGroup to get its backend name"


backend name of a processgroup created using distributed_c10d python API is tracked, but there is no good way to track name of a processgroup created using processGroup c++ API. In some cases, knowing backend name of a processGroup is useful, e,g., log the backend name, or write some codes that have dependency on the known backend.

Differential Revision: [D26059769](https://our.internmc.facebook.com/intern/diff/D26059769/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "add a c++ interface in processGroup to get its backend name"


backend name of a processgroup created using distributed_c10d python API is tracked, but there is no good way to track name of a processgroup created using processGroup c++ API. In some cases, knowing backend name of a processGroup is useful, e,g., log the backend name, or write some codes that have dependency on the known backend.

Differential Revision: [D26059769](https://our.internmc.facebook.com/intern/diff/D26059769/)

[ghstack-poisoned]
